File: ts/128/54.ts
function spiralOrder(matrix: number[][]): number[] {
  const m = matrix.length, n = matrix[0].length
  const ans: number[] = []
  let l = 0, t = 0, r = n - 1, b = m - 1
  while (l <= r && t <= b) {
    for (let i = l; i <= r && t <= b; i++) {
      ans.push(matrix[t][i])
    }
    t += 1

    for (let i = t; i <= b && l <= r; i++) {
      ans.push(matrix[i][r])
    }
    r -= 1

    for (let i = r; i >= l && t <= b; i--) {
      ans.push(matrix[b][i])
    }
    b -= 1

    for (let i = b; i >= t && l <= r; i--) {
      ans.push(matrix[i][l])
    }
    l += 1
  }
  return ans
};

File: py3/128/54.py
from typing import List

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        if not matrix or not matrix[0]: return ans

        m, n = len(matrix), len(matrix[0])
        t, r, b, l = 0, n - 1, m - 1, 0

        while l <= r and t <= b:
            for i in range(l, r + 1):
                ans.append(matrix[t][i])
            t += 1

            for i in range(t, b + 1):
                ans.append(matrix[i][r])
            r -= 1

            if t <= b:
                for i in range(r, l - 1, -1):
                    ans.append(matrix[b][i])
                b -= 1

            if l <= r:
                for i in range(b, t - 1, -1):
                    ans.append(matrix[i][l])
                l += 1

        return ans